src/lib/device-logs/backends/loki.ts

```diff
@@ -0,0 +1,188 @@
+	},
+	required: ['createdAt', 'timestamp', 'isSystem', 'isStdErr', 'message'],
+};
+
+interface LokiStream {
+	stream: {
+		device_id: string;
+		service_id: string;
+		is_system: string;
+		is_std_err: string;
+	};
+	values: Array<[string, string]>;
+}
+
+export class LokiBackend implements DeviceLogsBackend {
+	private subscriptions: EventEmitter;
+	private webSockets: Map<string, WebSocket>;
+
+	constructor() {
+		this.subscriptions = new EventEmitter();
+		this.webSockets = new Map();
+	}
+
+	public get available(): boolean {
+		// TODO: check /ready endpoint?
+		return true;
+	}
+
+	public async history(ctx: LogContext, count: number): Promise<DeviceLog[]> {
+		const [response, body] = await requestAsync({
+			url: `http://${LOKI_HOST}:${LOKI_PORT}/loki/api/v1/query_range`,
+			headers: {
+				'X-Scope-OrgID': ctx.belongs_to__application,
+			},
+			method: 'GET',
+			qs: {
+				query: this.getDeviceQuery(ctx),
+				limit: _.isFinite(count) ? count : 1000,
+			},
+			json: true,
+		});
+		if (response.statusCode === 200) {
+			return this.fromLokiStreamsToDeviceLogs(body.data.result);
+		} else {
+			throw new BadRequestError('Query failed');
+		}
+	}
+
+	public async publish(ctx: LogWriteContext, logs: DeviceLog[]): Promise<any> {
+		return requestAsync({
+			url: `http://${LOKI_HOST}:${LOKI_PORT}/loki/api/v1/push`,
+			headers: {
+				'X-Scope-OrgID': ctx.belongs_to__application,
+			},
+			method: 'POST',
+			json: {
+				streams: logs.map((log) => this.toLokiStream(ctx, log), this),
+			},
+		});
+	}
+
+	public subscribe(ctx: LogContext, subscription: Subscription) {
+		const key = this.getKey(ctx);
+		if (!this.webSockets.has(key)) {
+			const tailUri =
+				`ws://${LOKI_HOST}:${LOKI_PORT}/loki/api/v1/tail` +
+				`?query=${encodeURIComponent(this.getDeviceQuery(ctx))}&start=${String(
+					Date.now(),
+				)}000000`;
+			const ws = new WebSocket(tailUri, {
+				headers: { 'X-Scope-OrgID': ctx.belongs_to__application },
+			});
+			ws.on('message', (data: string) => {
+				this.fromLokiStreamsToDeviceLogs(JSON.parse(data).streams).map(
+					(log: DeviceLog) => {
+						this.subscriptions.emit(key, log);
+					},
+				);
+			});
+			ws.on('close', () => {
+				this.webSockets.delete(key);
+			});
+			this.webSockets.set(key, ws);
+		}
+		this.subscriptions.on(key, subscription);
+	}
+
+	public unsubscribe(ctx: LogContext, subscription: Subscription) {
+		const key = this.getKey(ctx);
+		this.subscriptions.removeListener(key, subscription);
+		if (!this.webSockets.has(key)) {
+			const ws = this.webSockets.get(key);
+			if (ws) ws.close();
+		}
+	}
+
+	private getDeviceQuery(ctx: LogContext) {
+		return `{device_id="${ctx.id}"}`;
+	}
+
+	private getKey(ctx: LogContext, suffix = 'logs') {
+		return `device:${ctx.id}:${suffix}`;
+	}
+
+	private fromLokiStreamsToDeviceLogs(streams: LokiStream[]): DeviceLog[] {
+		return _.concat(
+			[],
+			...streams.map((stream: LokiStream) => this.fromLokiStream(stream)),
+		);
```

Comment:
I think what you actually want here is
```suggestion
		return streams.flatMap((stream: LokiStream) => this.fromLokiStream(stream));
```
?
And you can probably simplify further with
```suggestion
		return streams.flatMap(this.fromLokiStream);
```

Reply:
Indeed much nicer! Thanks